1. 回调函数用途广泛 — 他们不仅仅可以用来控制函数的执行顺序和函数之间的数据传递，还可以根据环境的不同，将数据传递给不同的函数。

    有些回调函数是同步的，e.g.

   ```
   //使用 Array.prototype.forEach() 来遍历数组
   const gods = ['Apollo', 'Artemis', 'Ares', 'Zeus'];
   
   gods.forEach(function (eachName, index){
     console.log(index + '. ' + eachName);
   });
   ```

2. 回调地狱：e.g.

   ```
   setTimeout(function(){
       console.log("first");
       setTimeout(function(){
           console.log("second");
           setTimeout(function(){
               console.log("third");
               setTimeout(function(){
                   console.log("fourth");
               },2000);
           },2000);
       },2000);
   },2000);
   ```

3. Promise与旧式回调相比具有很多优点：

   - 可以使用多个then()操作将多个异步操作链接在一起，并将其中一个操作的结果作为输入传递给下一个操作。这种操作会使回调以混乱的“末日金字塔”告终。
   - `Promise`总是严格按照它们放置在事件队列中的顺序调用。
   - 所有的错误都由块末尾的一个.catch()块处理，而不是在“金字塔”的每一层单独处理。

4. 一个成功**resolved**的promise称为**fullfilled**（**实现**）。它返回一个值，可以通过将`.then()`块链接到promise链的末尾来访问该值。` .then()`块中的执行程序函数将包含promise的返回值。

5. 一个不成功**resolved**的promise被称为**rejected**（**拒绝**）了。它返回一个原因（**reason**），一条错误消息，说明为什么拒绝promise。可以通过将`.catch()`块链接到promise链的末尾来访问此原因。

6. `.then()`块返回一个新的promise,`.catch()`块来处理所有错误

7. 将`async`加入到函数声明中，返回的是`promise`。

8. `await`会暂停代码在该行之上，直到`promise`完成。

9. axios创建实例：`axios.create([config]) Create ([ config ])`

   ```
   const instance = axios.create({
     baseURL: 'https://some-domain.com/api/',
     timeout: 1000,
     headers: {'X-Custom-Header': 'foobar'}
   });
   ```

10. axios配置优先顺序:配置文件将被合并为一个优先顺序文件夹。顺序是在 lib/defaults.js 中找到的库默认值，然后是实例的默认属性，最后是请求的 config 参数。后者将优先于前者。

e.g.

```
// Create an instance using the config defaults provided by the library
// At this point the timeout config value is `0` as is the default for the library
const instance = axios.create();

// Override timeout default for the library
// Now all requests using this instance will wait 2.5 seconds before timing out
instance.defaults.timeout = 2500;

// Override timeout for this request as it's known to take a long time
instance.get('/longRequest', {
  timeout: 5000
});
```

11. 使用 `toJSON` 可以得到一个对象，其中包含有关 HTTP 错误的更多信息。

```
axios.get('/user/12345')
  .catch(function (error) {
    console.log(error.toJSON());
  });
```



